### PR TITLE
Add controller to cleanup stale credentials requests

### PR DESCRIPTION
**Pull request by akhil-rane**
This PR adds a controller to remove the credentials requests no longer required by the CCO

### CODE REVIEW

File: pkg/operator/cleanup/cleanup_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+			}
+			if err == nil && test.expectErr {
+				t.Errorf("Expected error but got none")
+			}
+
+			cr := getCR(fakeClient)
+			if test.expectDeletion {
+				assert.Nil(t, cr, "expected credentials request to be deleted")
+			} else {
+				assert.NotNil(t, cr, "expected credentials request to exist")
```
Comment (joelddiaz) [inline pkg/operator/cleanup/cleanup_controller_test.go:181]: Think you want a `require.NotNil` here b/c if it is nill, the code below will try to access `cr`.

File: pkg/operator/cleanup/cleanup_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+			}
+
+			for _, condition := range test.expectedConditions {
+				foundCondition := utils.FindCredentialsRequestCondition(cr.Status.Conditions, condition.conditionType)
+				require.NotNil(t, foundCondition, "unexpected unable to find condition")
+				assert.Exactly(t, condition.status, foundCondition.Status)
+				assert.Exactly(t, condition.reason, foundCondition.Reason)
+			}
```
Comment (joelddiaz) [inline pkg/operator/cleanup/cleanup_controller_test.go:189]: This loop seems like it should be in the NotNil scope right above.

File: pkg/operator/credentialsrequest/status.go [hunk 3]
```diff
@@ -22,8 +22,8 @@ import (
 )
 
 const (
-	reasonCredentialsFailing = "CredentialsFailing"
-	reasonReconciling        = "Reconciling"
+	reasonFailingOrStaleCredentials = "FailingOrStaleCredentials"
```
[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/operator/credentialsrequest/status.go:25]: I think it technically is. I'd do a separate reason if possible

File: pkg/operator/constants/constants.go [hunk 4]
```diff
@@ -145,4 +145,24 @@ var (
 	OvirtUpcomingSecrets = []types.NamespacedName{}
 	// VsphereUpcomingSecrets contains the list of known new vSphere credential secrets for the next version of OpenShift
 	VsphereUpcomingSecrets = []types.NamespacedName{}
+
+	// Add known stale credentials requests here
+
+	// AWSStaleCredentialsRequests contains the list of known stale AWS credentials requests for the next version of OpenShift
+	AWSStaleCredentialsRequests = []types.NamespacedName{
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/operator/constants/constants.go]: you are right, unified list should be enough

File: pkg/operator/cleanup/cleanup_controller.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+type ReconcileCredentialsRequest struct {
+	client.Client
+	Actuator     actuator.Actuator
+	platformType configv1.PlatformType
+}
+
+// ReconcileStaleCredentialsRequest reconciles a stale CredentialsRequest object
+type ReconcileStaleCredentialsRequest struct {
+	ReconcileCredentialsRequest credentialsrequest.ReconcileCredentialsRequest
+}
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/operator/cleanup/cleanup_controller.go]: that is just me who forgot to remove stale code.

File: pkg/operator/cleanup/cleanup_controller.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+		"secret": fmt.Sprintf("%s/%s", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name),
+	})
+
+	origCR := cr
+	cr = cr.DeepCopy()
+
+	if mode == operatorv1.CloudCredentialsModeManual {
+		logger.Warnf("operator set to disabled / manual mode, user needs to delete stale credentials")
+
+		msg := fmt.Sprintf("CredentialsRequest is no longer required")
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/operator/cleanup/cleanup_controller.go]: done

File: pkg/operator/cleanup/cleanup_controller.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+		if err != nil {
+			logger.WithError(err).Error("failed to update conditions")
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Delete stale credentials request
+	if cr.DeletionTimestamp == nil {
+		err = r.ReconcileCredentialsRequest.Client.Delete(context.TODO(), cr)
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/operator/cleanup/cleanup_controller.go:173]: We handle manual mode above. This will code is for non-manual mode.

File: pkg/operator/controller.go [hunk 8]
```diff
@@ -53,6 +54,7 @@ func init() {
 	AddToManagerFuncs = append(AddToManagerFuncs, awspodidentity.Add)
 	AddToManagerFuncs = append(AddToManagerFuncs, status.Add)
 	AddToManagerWithActuatorFuncs = append(AddToManagerWithActuatorFuncs, credentialsrequest.AddWithActuator)
+	AddToManagerWithActuatorFuncs = append(AddToManagerWithActuatorFuncs, cleanup.AddWithActuator)
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/operator/controller.go]: We do not need actuator in cleanup controller. Removed it

File: pkg/operator/cleanup/cleanup_controller_test.go (hunk 1 above)
Comment (akhil-rane) [reply pkg/operator/cleanup/cleanup_controller_test.go]: right! made the change

**Comment by akhil-rane** [comment]
@joelddiaz addressed your feedback PTAL.

File: pkg/operator/cleanup/cleanup_controller.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+		msg := fmt.Sprintf("CredentialsRequest is no longer required. Delete CR, Secret containing credentials, and associated platform/cloud resources")
+		reason := "CredentialsNoLongerRequired"
+		updateCheck := utils.UpdateConditionIfReasonOrMessageChange
+		status := corev1.ConditionTrue
+
+		cr.Status.Conditions = utils.SetCredentialsRequestCondition(cr.Status.Conditions, minterv1.StaleCredentials,
+			status, reason, msg, updateCheck)
+
+		err := r.ReconcileCredentialsRequest.UpdateStatus(origCR, cr, logger)
```
Comment (joelddiaz) [inline pkg/operator/cleanup/cleanup_controller.go:163]: Lets move this UpdateStatus() into a shared location so that ReconcileStaleCredentialsRequest does not need to embed ReconcileCredentialsRequest

File: pkg/operator/credentialsrequest/credentialsrequest_controller.go [hunk 10]
```diff
@@ -90,7 +90,7 @@ func newReconciler(mgr manager.Manager, actuator actuator.Actuator, platType con
 	r := &ReconcileCredentialsRequest{
 		Client:       mgr.GetClient(),
 		Actuator:     actuator,
-		platformType: platType,
+		PlatformType: platType,
```
Comment (joelddiaz) [inline pkg/operator/credentialsrequest/credentialsrequest_controller.go:93]: can we undo the changes of platformType to PlatformType once ReconcileStaleCredentialsRequest doesn't embed ReconcileCredentialsRequest?

**Review by joelddiaz — COMMENTED** [review]
this looks good. the last thing i'd like to see is reverting the exporting of the `platformType` field in the struct.

File: pkg/operator/cleanup/cleanup_controller_test.go [hunk 11]
```diff
@@ -186,21 +177,6 @@ func TestStaleCredentialsRequestReconcile(t *testing.T) {
...
-
-				for _, expectedCondition := range test.expectedCOConditions {
-					foundCondition := utils.FindClusterOperatorCondition(currentConditions, expectedCondition.conditionType)
-					require.NotNil(t, foundCondition)
-					assert.Equal(t, string(expectedCondition.status), string(foundCondition.Status), "condition %s had unexpected status", expectedCondition.conditionType)
-					if expectedCondition.reason != "" {
-						assert.Exactly(t, expectedCondition.reason, foundCondition.Reason)
-					}
-				}
-			}
```
Comment (akhil-rane) [inline pkg/operator/cleanup/cleanup_controller_test.go]: Removed Cluster Operator conditions checking as these are set by ReconcileCredentialsRequest status handler instead.

File: pkg/operator/credentialsrequest/status.go [hunk 12]
```diff
@@ -152,7 +152,7 @@ func computeStatusConditions(
 			degradedCondition.Reason = reasonStaleCredentials
 			degradedCondition.Message = fmt.Sprintf(
 				"%d of %d credentials requests are stale and should be deleted.",
-				failingCredRequests, len(validCredRequests))
+				staleCredRequests, len(validCredRequests))
```
Comment (akhil-rane) [inline pkg/operator/credentialsrequest/status.go:155]: fixed error in reporting count of stale cred requests.

**Review by akhil-rane — COMMENTED** [review]
@joelddiaz Addressed your feedback and some minor fixes. PTAL.

**Comment by joelddiaz** [comment]
I still see the ReconcileCredentialsRequest struct embedded inside the ReconcileStaleCredentialsRequest. Why is this needed?

**Comment by akhil-rane** [comment]
@joelddiaz you were ReconcileCredentialsRequest is not required here. Updated the PR.

**Review by joelddiaz — COMMENTED** [review]
/lgtm

**Comment by sdodson** [comment]
/close
to be re-opened
